### PR TITLE
HTCONDOR-872 submit side support for container universe

### PR DESCRIPTION
**Pull request by GregThain**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [ ] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_utils/submit_utils.cpp [hunk 1]
```diff
@@ -7083,6 +7161,46 @@ int SubmitHash::process_vm_input_files(StringList & input_files, long long * acc
 	return count;
 }
 
+int SubmitHash::process_container_input_files(StringList & input_files, long long * accumulate_size_kb) {
+	auto_free_ptr transfer_container(submit_param(SUBMIT_KEY_TransferContainer, ATTR_TRANSFER_CONTAINER));
+	auto_free_ptr container_image(submit_param(SUBMIT_KEY_ContainerImage, ATTR_CONTAINER_IMAGE));
+
+	// User said don't transfer the container
+	if (transfer_container && strcasecmp(transfer_container.ptr(), "false")) {
```
Comment (johnkn) [inline src/condor_utils/submit_utils.cpp:7169]: should use submit_param_bool here, not compare a string to false.

File: src/condor_utils/submit_utils.cpp [hunk 2]
```diff
@@ -7083,6 +7161,46 @@ int SubmitHash::process_vm_input_files(StringList & input_files, long long * acc
...
+		return 0;
+	}
+
+	// don't xfer if on known shared fs
+	if (container_image.ptr()) {
```
Comment (johnkn) [inline src/condor_utils/submit_utils.cpp:7174]: just `if (container_image)` here

File: src/condor_utils/submit_utils.cpp [hunk 3]
```diff
@@ -7113,6 +7231,28 @@ int SubmitHash::process_input_file_list(StringList * input_list, long long * acc
 	return 0;
 }
 
+SubmitHash::ContainerImageType 
+SubmitHash::image_type_from_string(const std::string &image) const {
+	if (starts_with(image, "docker:")) {
```
Comment (johnkn) [inline src/condor_utils/submit_utils.cpp:7236]: is case sensitive test correct here?

Comment (GregThain) [reply src/condor_utils/submit_utils.cpp:7236]: yes, the docker prefix must be lower case

File: src/condor_utils/submit_utils.cpp (hunk 2 above)
Comment (GregThain) [reply src/condor_utils/submit_utils.cpp]: and done.

**Review by johnkn — APPROVED** [review]
Looks good

**Pull request merged**